src/index/CompressedRelation.cpp

```diff
@@ -312,80 +341,140 @@ bool CompressedRelationReader::FilterDuplicatesAndGraphs::canBlockBeSkipped(
 }
 
 // _____________________________________________________________________________
-CompressedRelationReader::IdTableGenerator CompressedRelationReader::lazyScan(
+CompressedRelationReader::IdTableGeneratorInputRange
+CompressedRelationReader::lazyScan(
     ScanSpecification scanSpec,
     std::vector<CompressedBlockMetadata> relevantBlockMetadata,
     ColumnIndices additionalColumns, CancellationHandle cancellationHandle,
     const LocatedTriplesPerBlock& locatedTriplesPerBlock,
     LimitOffsetClause limitOffset) const {
   AD_CONTRACT_CHECK(cancellationHandle);
 
-  // We will modify `limitOffset` as we go. We make a copy of the original
-  // value for some sanity checks at the end of the function.
-  const auto originalLimit = limitOffset;
-
   if (relevantBlockMetadata.empty()) {
-    co_return;
+    return IdTableGeneratorInputRange{};
   }
 
-  // Some preparation.
-  auto [beginBlockMetadata, endBlockMetadata] =
-      getBeginAndEnd(relevantBlockMetadata);
-  LazyScanMetadata& details = co_await cppcoro::getDetails;
-  size_t numBlocksTotal = endBlockMetadata - beginBlockMetadata;
-  auto config =
-      getScanConfig(scanSpec, additionalColumns, locatedTriplesPerBlock);
+  struct Generator : ad_utility::InputRangeFromGet<IdTable, LazyScanMetadata> {
+    ScanSpecification scanSpec;
+    std::vector<CompressedBlockMetadata> relevantBlockMetadata;
+    ColumnIndices additionalColumns;
+    CancellationHandle cancellationHandle;
+    const LocatedTriplesPerBlock& locatedTriplesPerBlock;
+    LimitOffsetClause limitOffset;
+    ad_utility::InputRangeTypeErased<IdTable, LazyScanMetadata>
+        blockGenerator{};
+    bool firstBlockYielded{false};
+    bool middleBlocksYielded{false};
+    bool lastBlockYielded{false};
+    bool generatorCreated{false};
```

Comment:
It looks like you are writing a good old manual state machine. Then this is probably an enum, and not for (not so independent) boolean flags.

Reply:
yes, I will change this to enum